### PR TITLE
Add inserting and removing bytes locally

### PR DESCRIPTION
**Pull request by catsuryuu**

### CODE REVIEW

File: src/util/edit.cc [hunk 1]
```diff
@@ -214,6 +250,40 @@ data::BinData EditEngine::getDataFromEditNode(const EditNode& edit_node,
...
+  assert(pos_next_it != address_mapping_.cbegin());
+
+  QMap<size_t, EditNode> new_address_mapping;
+  for (auto it = address_mapping_.cbegin(); it != pos_next_it; ++it) {
+    new_address_mapping.insert(it.key(), it.value());
+  }
+
+  size_t insert_pos = pos + old_size;
+  auto insert_pos_next_it =
+      old_size > 0 ? address_mapping_.upperBound(insert_pos) : pos_next_it;
```
Comment (mkow) [inline src/util/edit.cc]: Please don't micro-optimize like this, this function is O(n) and this saves just one O(lg n) operation in a rare corner-case ;)
Just write:
```C++
auto insert_pos_next_it = address_mapping_.upperBound(insert_pos);
```

File: src/util/edit.cc [hunk 2]
```diff
@@ -214,6 +250,40 @@ data::BinData EditEngine::getDataFromEditNode(const EditNode& edit_node,
...
+
+  auto pos_next_it = address_mapping_.upperBound(pos);
+  assert(pos_next_it != address_mapping_.cbegin());
+
+  QMap<size_t, EditNode> new_address_mapping;
+  for (auto it = address_mapping_.cbegin(); it != pos_next_it; ++it) {
+    new_address_mapping.insert(it.key(), it.value());
+  }
+
+  size_t insert_pos = pos + old_size;
```
Comment (mkow) [inline src/util/edit.cc]: Could you add an explanation in a comment why do you pick this certain value? (`pos + old_size`)

File: src/util/edit.cc [hunk 3]
```diff
@@ -214,6 +250,40 @@ data::BinData EditEngine::getDataFromEditNode(const EditNode& edit_node,
...
+  --insert_pos_it;
+
+  new_address_mapping.insert(
+      insert_pos + offset,
+      EditNode(insert_pos_it->fragment_,
+               insert_pos_it->offset_ + (insert_pos - insert_pos_it.key())));
+
+  for (auto it = insert_pos_next_it; it != address_mapping_.cend(); ++it) {
+    new_address_mapping.insert(it.key() + offset, it.value());
+    // it can overwrite the `*pos_it` node and that's OK
```
Comment (mkow) [inline src/util/edit.cc]: ditto

File: src/util/edit.cc [hunk 4]
```diff
@@ -214,6 +250,40 @@ data::BinData EditEngine::getDataFromEditNode(const EditNode& edit_node,
...
+  }
+
+  address_mapping_ = new_address_mapping;
```
Comment (mkow) [inline src/util/edit.cc]: Use std::move here to avoid copying.

File: src/ui/hexedit.cc [hunk 5]
```diff
@@ -83,6 +83,10 @@ void HexEdit::recalculateValues() {
 
   horizontalScrollBar()->setRange(0, lineWidth_ - viewport()->width());
   startPosX_ = horizontalScrollBar()->value();
+
+  if (selectionEnd() > dataBytesCount_) {
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply src/ui/hexedit.cc]: It's also possible after 'discard' operation.

File: src/ui/hexedit.cc [hunk 6]
```diff
@@ -268,6 +283,7 @@ HexEdit::HexEdit(FileBlobModel* dataModel, QItemSelectionModel* selectionModel,
 
   menu_.addSeparator();
 
+  menu_.addAction(deleteSelectionAction_);
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply src/ui/hexedit.cc]: Later I'll do a big reordering of all operations.

File: src/ui/hexeditwidget.cc [hunk 7]
```diff
@@ -217,6 +217,23 @@ void HexEditWidget::createActions() {
   auto_resize_act_->setDefaultWidget(auto_resize_checkbox_);
   connect(auto_resize_checkbox_, &QCheckBox::toggled,
           [this](bool toggled) { hex_edit_->setAutoBytesPerRow(toggled); });
+
+  change_edit_mode_button_ = new QPushButton("override");
```
[the remark this answers is not in the document]
Comment (mkow) [reply src/ui/hexeditwidget.cc]: `Override` -> `Overwrite`

File: include/ui/hexedit.h [hunk 8]
```diff
@@ -58,6 +58,8 @@ class HexEdit : public QAbstractScrollArea {
   void processEditEvent(QKeyEvent* event);
   uint64_t byteValue(qint64 pos) const;
   void setBytesValues(qint64 pos, const data::BinData& new_data);
+  bool isInsertMode() { return insert_mode_; }
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply include/ui/hexedit.h]: DONE

File: include/ui/hexedit.h [hunk 9]
```diff
@@ -151,6 +153,7 @@ class HexEdit : public QAbstractScrollArea {
   WindowArea current_area_;
   qint64 cursor_pos_in_byte_;
   bool cursor_visible_;
+  bool insert_mode_;
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply include/ui/hexedit.h]: DONE

File: src/ui/hexeditwidget.cc [hunk 10]
```diff
@@ -217,6 +217,23 @@ void HexEditWidget::createActions() {
...
+  change_edit_mode_act_ = ShortcutsModel::getShortcutsModel()->createQAction(
+      util::settings::shortcuts::CHANGE_EDIT_MODE, this,
+      Qt::WidgetWithChildrenShortcut);
+  connect(change_edit_mode_act_, &QAction::triggered, [this]() {
+    if (hex_edit_->isInsertMode()) {
+      hex_edit_->setInsertMode(false);
+      change_edit_mode_button_->setText("override");
+    } else {
+      hex_edit_->setInsertMode(true);
+      change_edit_mode_button_->setText("insert");
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply src/ui/hexeditwidget.cc]: DONE

File: src/ui/hexeditwidget.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (catsuryuu) [reply src/ui/hexeditwidget.cc]: DONE

File: src/util/edit.cc [hunk 11]
```diff
@@ -80,6 +80,42 @@ void EditEngine::changeBytes(size_t pos, const data::BinData& bytes,
...
+  if (add_to_history) {
+    // TODO(catsuryuu)
+  }
+
+  has_changes_ = true;
+
+  remap(pos, 0, bytes.size());
+
+  auto it = address_mapping_.insert(pos, EditNode(bytes));
+  // it can overwrite the node with `pos` key and that's OK
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply src/util/edit.cc]: `This`. Fixed.

File: src/util/edit.cc [hunk 12]
```diff
@@ -214,6 +250,40 @@ data::BinData EditEngine::getDataFromEditNode(const EditNode& edit_node,
   return edit_node.fragment_->data(edit_node.offset_ + offset, size);
 }
 
+void EditEngine::remap(size_t pos, size_t old_size, size_t new_size) {
+  // This variable can be overflowed for `old_size` > `new_size`,
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: src/util/edit.cc (hunk 1 above)
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: src/util/edit.cc (hunk 3 above)
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: src/util/edit.cc (hunk 4 above)
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: src/ui/hexeditwidget.cc [hunk 13]
```diff
@@ -217,6 +217,23 @@ void HexEditWidget::createActions() {
...
+  connect(change_edit_mode_button_, &QPushButton::clicked,
+          [this]() { change_edit_mode_act_->trigger(); });
+
+  change_edit_mode_act_ = ShortcutsModel::getShortcutsModel()->createQAction(
+      util::settings::shortcuts::CHANGE_EDIT_MODE, this,
+      Qt::WidgetWithChildrenShortcut);
+  connect(change_edit_mode_act_, &QAction::triggered, [this]() {
+    if (hex_edit_->isInsertMode()) {
+      hex_edit_->setInsertMode(false);
+      change_edit_mode_button_->setText("override");
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply src/ui/hexeditwidget.cc]: DONE

File: src/util/edit.cc (hunk 2 above)
Comment (catsuryuu) [reply src/util/edit.cc]: Done

File: src/ui/hexeditwidget.cc [hunk 14]
```diff
@@ -218,22 +218,25 @@ void HexEditWidget::createActions() {
...
-    if (hex_edit_->isInsertMode()) {
-      hex_edit_->setInsertMode(false);
-      change_edit_mode_button_->setText("override");
-    } else {
-      hex_edit_->setInsertMode(true);
-      change_edit_mode_button_->setText("insert");
-    }
-  });
+  connect(change_edit_mode_act_, &QAction::triggered,
+          [this, overwrite_label_]() {
```
Comment (mkow) [inline src/ui/hexeditwidget.cc]: Should be `&overwrite_label_`. A reference captured by value causes copying of the original object. We want a reference to `QStringLiteral("Overwrite")` (which is a static object, so this is ok).

**Pull request merged**